### PR TITLE
Use annotations instead of delete CRDs when terminate execution

### PR DESCRIPTION
**Pull request by andresgomezfrr**
## Tracking issue
Closes https://github.com/flyteorg/flyte/issues/5786
Alternative to: https://github.com/flyteorg/flyte/pull/5788

## Why are the changes needed?

To make sure that the propeller can handle the CRD before they are deleted if a terminated execution is called very fast.

**New behavior:**
Instead of propagating the terminate execution to the propeller deleting the CRD, the admin now updates a new aborted annotation, the propeller will check this annotation instead of the `deletedTimestamp`. If the annotation is there the propeller will send the `ABORT` event to the admin and remove the finalizer from the CRD. The propeller garbage-collector will clean up this CRD when it is executed.

## What changes were proposed in this pull request?
Unit test + manual test in a staging environment with real flyte execution.

## Check all the applicable boxes 

- [ ] I updated the documentation accordingly.
- [x] All new and existing tests passed.
- [x] All commits are signed-off.

### CODE REVIEW

File: flytepropeller/pkg/controller/handler.go [hunk 1]
```diff
@@ -103,7 +110,8 @@ func (p *Propeller) TryMutateWorkflow(ctx context.Context, originalW *v1alpha1.F
 	ctx = contextutils.WithResourceVersion(ctx, mutableW.GetResourceVersion())
 
 	maxRetries := uint32(p.cfg.MaxWorkflowRetries)
-	if IsDeleted(mutableW) || (mutableW.Status.FailedAttempts > maxRetries) {
+
+	if IsAborted(mutableW) || (mutableW.Status.FailedAttempts > maxRetries) {
```
Comment (pvditt) [inline flytepropeller/pkg/controller/handler.go:114]: we could still have the case that the CRD gets deleted by some other process and should still account for that here.

File: flytepropeller/pkg/controller/handler.go [hunk 2]
```diff
@@ -89,6 +91,11 @@ func SetDefinitionVersionIfEmpty(wf *v1alpha1.FlyteWorkflow, version v1alpha1.Wo
 	}
 }
 
+func IsAborted(w *v1alpha1.FlyteWorkflow) bool {
```
Comment (pvditt) [inline flytepropeller/pkg/controller/handler.go:94]: can this be moved to the same place IsDeleted is defined?

File: flyteadmin/pkg/workflowengine/impl/k8s_executor.go [hunk 3]
```diff
@@ -94,13 +93,27 @@ func (e K8sWorkflowExecutor) Abort(ctx context.Context, data interfaces.AbortDat
...
+	} else if k8_api_err.IsNotFound(err) {
+		logger.Infof(ctx, "Trying to abort an execution [%+v] that is not found in cluster: %s, skipping...", data.ExecutionID, target.ID)
+		return nil
+	}
+
+	if w.ObjectMeta.Annotations == nil {
+		w.ObjectMeta.Annotations = make(map[string]string)
+	}
+
+	w.ObjectMeta.Annotations[AbortedWorkflowAnnotation] = "true"
```
Comment (pvditt) [inline flyteadmin/pkg/workflowengine/impl/k8s_executor.go:109]: concern here is if there's a case that this delete is required to delete dangling CRDs (not picked up by propeller)

**Comment by andresgomezfrr** [comment]
I think that the propeller GC should remove that workflows, right?

We checked in our staging and they were cleaned up

File: flytepropeller/pkg/controller/handler.go (hunk 1 above)
Comment (andresgomezfrr) [reply flytepropeller/pkg/controller/handler.go:114]: mmmmh could be? Looking inside the if, the function there looks like is only for aborting case

File: flytepropeller/pkg/controller/handler.go (hunk 2 above)
Comment (andresgomezfrr) [reply flytepropeller/pkg/controller/handler.go:94]: Sure I will do it, tomorrow

File: flyteadmin/pkg/workflowengine/impl/k8s_executor.go (hunk 3 above)
Comment (andresgomezfrr) [reply flyteadmin/pkg/workflowengine/impl/k8s_executor.go:109]: I guess.. that the dangling CRDs should be removed by propeller GC, right?

File: flytepropeller/pkg/controller/handler.go (hunk 2 above)
Comment (andresgomezfrr) [reply flytepropeller/pkg/controller/handler.go]: done!

File: flyteadmin/pkg/workflowengine/impl/k8s_executor.go (hunk 3 above)
Comment (andresgomezfrr) [reply flyteadmin/pkg/workflowengine/impl/k8s_executor.go:109]: It is the same case as completed workflows/failed workflows.

File: flytepropeller/pkg/controller/handler.go (hunk 1 above)
Comment (andresgomezfrr) [reply flytepropeller/pkg/controller/handler.go]: I added back the isDeleted(...) to manage this case here too, now it is backward compatible, wdyt?

File: flyteadmin/pkg/workflowengine/impl/k8s_executor.go (hunk 3 above)
Comment (andresgomezfrr) [reply flyteadmin/pkg/workflowengine/impl/k8s_executor.go:109]: >concern here is if there's a case that this delete is required to delete dangling CRDs (not picked up by propeller) 

If the concern is about deleting CRDs if the propeller is down.. what do you think about having a flag in the terminate to force the delete from the admin? Similar to kubectl delete --force which force the deletion of a CRD.

**Pull request closed without merging**